### DISCUSSION TITLE
关于匹配规则Rule和权限控制Permission的问题

### DISCUSSION THREAD

**Discussion by Redmomn** [nonebot/discussions#317] [问答]
请问一下`Rule`和`Permission`的区别在哪，(下面的疑问都基于当我用`Rule`来进行权限控制的情况下)
<img width="640" alt="image" src="https://user-images.githubusercontent.com/109732988/217827186-9c7af354-72df-49a3-8b50-3b51867d5c9e.png">
我对这句话的理解是，`Rule`只会响应第一次，在以后的规则匹配中失效，但是我测试之后发现`Rule`对每个注入它的事件响应器都会进行规则匹配，显然我的理解有问题，所以这句话应该怎么理解？
<img width="640" alt="image" src="https://user-images.githubusercontent.com/109732988/217828406-5adffe46-ea36-4b3a-93de-a705989936a3.png">
另外在这句话中，会话状态更新的具体定义是什么？
如果可以，麻烦给一个示例演示一下这两个的区别

**Comment by yanyongyu** [comment]
你截图中的说明都是以会话为scope的，也就是对单个事件响应器来说，而不是多个事件响应器。

举个简单的例子

```python
from nonebot import on_message

matcher = on_message(rule=some_rule, permission=some_permission)

@matcher.got("key", prompt="key?")
async def _():
    ...
```

在上面代码中，我们机器人和用户进行了一轮问答，即：用户发送消息，机器人询问key，用户回复key。

rule只有在首次用户发送消息时执行，permission会在用户发送消息和用户回复key时都执行一次。具体的操作就是在用户首次发送消息之后会丢掉rule，这样之后的对话中就不再会有rule执行；而permission会进行可控的更新，在会话中每次用户发送消息（或者事件）都执行

**Reply by Redmomn** [reply]
我之前的理解是，假如有一个装饰器A用于消息响应，A上使用了匹配规则B(即一个Rule),权限控制C(Permission)，在A第一次被触发时，BC都会运行，在A以后被触发的时候，只有C会运行，B会失效 

```cmd
@A.handle(rule = Rule(B), Permission = C )
async def _():
    假设这里有一个消息响应D
```
我不知道nonebot有没有这样的写法，我在这里只是想表达这样一个意思

现在我可不可以这样理解，在A被触发之后，BC都会运行，但是在接下来的D中只会运行C

**Reply by yanyongyu** [reply]
在A首次响应时执行了B和C，此时就进入了A的会话X，在会话X中进行的任何消息交互都只会执行C
